core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala

```diff
@@ -133,20 +160,200 @@ private[op] class OpMultiClassificationEvaluator
         topKs = $(topKs)
       )
 
+      val rddCm = dataUse.select(col(labelColName), col(predictionColName), col(probabilityColName)).rdd.map{
+        case Row(label: Double, pred: Double, prob: DenseVector) => (label, pred, prob.toArray)
+      }
+      val topKConfusionMatrix = calculateTopKConfusionMatrixByThreshold(rddCm)
+
+      val topNMisclassifications = calculateMisClassificationsTopN(rddCm.map{ case (label, pred, _) => (label, pred)})
+
       val metrics = MultiClassificationMetrics(
         Precision = precision,
         Recall = recall,
         F1 = f1,
         Error = error,
         ThresholdMetrics = thresholdMetrics,
-        TopKMetrics = topKMetrics
+        TopKMetrics = topKMetrics,
+        TopKConfusionMatrix = topKConfusionMatrix,
+        TopNMisclassification = topNMisclassifications
       )
 
       log.info("Evaluated metrics: {}", metrics.toString)
       metrics
     }
   }
 
+/**
+ * function to find the lower bound of the confidence threshold bin in which the probability score falls.
+ * E.g. if thresholds(i) <= probability < thresholds(i+1), return thresholds(i)
+ *
+ * @param probability the probability score of the predicted class
+ * @param sortedThresholds thresholds sorted in ascending order
+ * @return lower bound of the
+ */
+  def findConfidenceThresholdBin(probability: Double, sortedThresholds: Seq[Double]): Double = {
+
+    if (probability < sortedThresholds(0)) {0.0}
```

Comment:
We shouldn't need to implement our own binary search here. If you cast `sortedThresholds` to an `indexedSeq`, then you can use `scala.collection.Searching` here: https://github.com/scala/scala/blob/v2.11.0-M3/src/library/scala/collection/Searching.scala#L61-L99

It can find both exact matches and insertion points for elements not in the thresholds.